File: face/mouth_client.py
"""HTTP client for MOUTH TTS service.

MOUTH provides text-to-speech audio synthesis.
FACE polls GET /audio/next to retrieve completed audio chunks.
"""

import httpx

from config import config


class MouthClient:
    """Client for fetching audio from MOUTH TTS service."""

    def __init__(self, base_url: str = None):
        self.base_url = base_url or config.mouth_url

    def get_next_audio(self) -> tuple[bytes | None, dict]:
        """Fetch next audio chunk from TTS service.

        Returns:
            Tuple of (audio_bytes or None, metadata dict)
        """
        try:
            with httpx.Client(timeout=config.timeouts.connect) as client:
                response = client.get(f"{self.base_url}/audio/next")

                metadata = {
                    "pending": int(response.headers.get("X-Pending-Count", 0)),
                    "completed": int(response.headers.get("X-Completed-Count", 0)),
                }

                if response.status_code == 200:
                    return response.content, metadata
                elif response.status_code == 204:
                    return None, metadata
                else:
                    return None, metadata

        except Exception as e:
            return None, {"error": str(e)}

    def health_check(self) -> bool:
        """Check if MOUTH TTS service is available."""
        try:
            with httpx.Client(timeout=config.timeouts.health_check) as client:
                response = client.get(f"{self.base_url}/health")
                return response.status_code == 200
        except Exception:
            return False

File: face-android/app/src/main/java/com/homelab/cici/api/MouthClient.java
package com.homelab.cici.api;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class MouthClient {
    private final OkHttpClient client;
    private String baseUrl;

    public MouthClient(String host, int port) {
        this.baseUrl = "http://" + host + ":" + port;
        this.client = new OkHttpClient.Builder()
                .connectTimeout(5, TimeUnit.SECONDS)
                .readTimeout(5, TimeUnit.SECONDS)
                .build();
    }

    public void updateEndpoint(String host, int port) {
        this.baseUrl = "http://" + host + ":" + port;
    }

    public static class AudioResult {
        public final byte[] audioData;
        public final int pendingCount;
        public final int completedCount;

        public AudioResult(byte[] audioData, int pendingCount, int completedCount) {
            this.audioData = audioData;
            this.pendingCount = pendingCount;
            this.completedCount = completedCount;
        }
    }

    /**
     * Fetch next audio chunk from MOUTH. Returns null if no audio available (204).
     */
    public AudioResult getNextAudio() throws IOException {
        Request request = new Request.Builder()
                .url(baseUrl + "/audio/next")
                .build();

        try (Response response = client.newCall(request).execute()) {
            if (response.code() == 204) {
                return null;
            }
            if (!response.isSuccessful()) {
                throw new IOException("MOUTH returned " + response.code());
            }

            byte[] data = response.body().bytes();
            int pending = parseIntHeader(response, "X-Pending-Count", 0);
            int completed = parseIntHeader(response, "X-Completed-Count", 0);

            return new AudioResult(data, pending, completed);
        }
    }

    public boolean healthCheck() {
        Request request = new Request.Builder()
                .url(baseUrl + "/health")
                .build();
        try (Response response = client.newCall(request).execute()) {
            return response.isSuccessful();
        } catch (IOException e) {
            return false;
        }
    }

    private int parseIntHeader(Response response, String name, int defaultValue) {
        String val = response.header(name);
        if (val == null) return defaultValue;
        try {
            return Integer.parseInt(val);
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }
}
